### PR TITLE
k8s: add K8s nodes equalness and missing function

### PR DESCRIPTION
**Pull request by aanm**
With this commit, Cilium will no longer send unnecessary AddFunc events
and running useless operations for nodes that have non-relevant
changes.
    
Signed-off-by: André Martins <andre@cilium.io>

Read per commit basis.

```release-note
Stop triggering unnecessary ipcache setup for every Kubernetes node watch event.
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5659)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: daemon/k8s_watcher.go [hunk 1]
```diff
@@ -2174,3 +2172,23 @@ func (d *Daemon) deleteK8sNodeV1(k8sNode *v1.Node) {
...
 	}
 }
+
+// missingK8sNodeV1 checks if all nodes from the possible missing nodes have
+// their CiliumHostIP missing from the ipcache.
+func (d *Daemon) missingK8sNodeV1(m versioned.Map) versioned.Map {
+	missing := versioned.NewMap()
+	for k, v := range m {
+		n := v.Data.(*v1.Node)
+		ciliumHostIP := n.GetAnnotations()[annotation.CiliumHostIP]
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: Can we just use `node.GetNode()`?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: daemon/k8s_watcher.go (hunk 1 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: @tgraf I'm a little lost here, how would it help?

Comment (tgraf) [reply daemon/k8s_watcher.go]: Well, it is a lookup function into the list of nodes that Cilium is aware of. the current implementation relies on the annotation of the node resource.

Comment (aanm) [reply daemon/k8s_watcher.go]: but `m` is the list of nodes fetched from kube-apiserver. `node.GetNode()` is the list of nodes that Cilium is aware of based on other sources, like kv-store.

Comment (tgraf) [reply daemon/k8s_watcher.go]: The ipcache is also populated via the kvstore. Maybe I misunderstand the intention of the code here.

Comment (aanm) [reply daemon/k8s_watcher.go]: The intention is simple, either from the kv-store or kubernetes notification events, we only want the ipcache to contain `ciliumHostIP`. That's what we are checking here.

Comment (tgraf) [reply daemon/k8s_watcher.go]: OK. Please explain to me why we add all of the complexity to lookup via `ciliumHostIP` if we have a perfect API to lookup via the provided hostname. This implementation is vulnerable to potentially miss identifying if the nodename <-> ciliumHostIP changes as the lookup via ciliumHostIP would still succeed whereas a proper node name lookup followed by key attribute comparison would catch all of that.

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Pull request merged**